java/src/jni/h5util.c

```diff
@@ -697,9 +694,70 @@ h5str_sprint_reference(JNIEnv *env, h5str_t *out_str, hid_t container, void *ref
     }
     if (!h5str_append(out_str, "\""))
         CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+
+    ret_value = SUCCEED;
 done:
     if (ref_name)
         HDfree(ref_name);
+
+    return ret_value;
+} /* h5str_sprint_reference */
+
+int
+h5str_region_dataset(JNIEnv *env, h5str_t *out_str, H5R_ref_t *ref_vp, int expand_data)
+{
+    hid_t        new_obj_id  = H5I_INVALID_HID;
+    hid_t        new_obj_sid = H5I_INVALID_HID;
+    H5S_sel_type region_type;
+
+    int ret_value = FAIL;
+
+    if ((new_obj_id = H5Ropen_object(ref_vp, H5P_DEFAULT, H5P_DEFAULT)) >= 0) {
```

Comment:
It would be good if this was re-written to not have all the nesting. It makes it difficult to see that the code appears to be ignoring all library errors except for the one from H5Ropen_region, when we should probably be handling all errors nicely.

Reply:
Interesting idea, because I was capturing all the errors thinking that H5Ropen_object would return an invalid id like it  should! Something changed and the error doesn't happen until later.

Reply:
It can be annoying at times because when you need to exit a function and "not do anything" after a pre-condition check, you're usually forced to rely on goto, as well as manual setting of an error for the return value. On the other hand, it's done everywhere in HDF5 and also makes it so we don't end up with several levels of unnecessary nesting. It's the usual trade-off war.

Reply:
I can agree with that, especially since the close are now at the end. I will rework the nesting.

What about the check for expand_data - move that into the select ifs and only have one select block - six of one, half-dozen of the other to me.

Reply:
I think either way is fine really. The most ideal would probably be to have a switch around the select types and then handle expand_data inside the case statements, but it's not functionally any different.